Adding support of common security cipher module for encryption and decryption of a passkey

nmoray opened the pull request:
#### Why I did it
This module is created to handle the passkey encryption, decryption and the cipher storage. It's a common module which will be used for feature like TACACS, RADIUS, LDAP etc.
This implementation is aligned with [HLD](https://github.com/sonic-net/SONiC/pull/1471)

#### How I did it
This module will expose following APIs.
1. Passkey encryption for a given feature
2. Passkey decryption for a given feature
3. Check whether encryption feature is enabled for a given feature

#### How to verify it
`
Cipher / password used for the encryption:
root@sonic:/tmp# cat /etc/cipher_pass 
#Auto generated file for storing the encryption passwords
TACPLUS : TEST1
RADIUS : TEST2
LDAP : TEST3


Is encryption enabled for TACACS: False
Encrypted passkey for Feature: TACPLUS - U2FsdGVkX1/frdwl4GGD7bTKyzLi+lr2K76v0IECzkQ=
Passkey post decryption:TACPLUS - passkey1
Encrypted passkey for Feature: RADIUS - U2FsdGVkX1/fdiBo3RWWxIIPFJYCy1CF/ZQeLt8N96Q=
Passkey post decryption: RADIUS - passkey2
Encrypted passkey for Feature: LDAP - U2FsdGVkX1/o0UkHtWgOjr46UzLQRhXKAHngctey9TE=
Passkey post decryption: LDAP - passkey3
`
#### Which release branch to backport (provide reason below if selected)
<!--
- Note we only backport fixes to a release branch, *not* features!
- Please also provide a reason for the backporting below.
- e.g.
- [x] 202006
-->

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [x] 202111
- [ ] 202205
- [x] 202211
- [ ] 202305

#### Tested branch (Please provide the tested image version)
Master

#### Link to config_db schema for YANG module changes
[Schema](https://github.com/sonic-net/sonic-buildimage/blob/master/src/sonic-yang-models/doc/Configuration.md#tacplus-server)

venkatmahalingam commented on src/sonic-py-common/sonic_py_common/security_cipher.py:
```diff
@@ -0,0 +1,120 @@
+            self._config_db.connect()
+            # Note: Kept 1st index NA intentionally to map it with the cipher_pass file
+            # contents. The file has a comment at the 1st row / line
+            self._feature_list = ["NA", "TACPLUS", "RADIUS", "LDAP"]
```
Do we need a master key per feature? I thought we're using one master key for all features.

nmoray replied:
I guess it is better to have different Master keys for different feature. This way, there will not be any inter-dependency between them.

venkatmahalingam replied:
Let's not complicate things for the user, just follow well established practices (e.g a master key for all features) from the popular NOS if possible, my 2 cents.

nmoray replied:
With this approach, we can further extend this implementation for other modules too not just TACPLUS, RADIUS and LDAP. Additionally, it is upto the user if he / she needs to use different keys or can use same keys as well.

nmoray replied:
It will give flexibility to the user, in my opinion.

nmoray replied:
If we go with single key, it will be having huge dependancy in case of changing that key. User needs to change the encrypted passkey in CONFIG_DB for all the features.

venkatmahalingam replied:
I disagree, IMO, we should just look at the behavior of other popular NOS that's been there for many years for now, if required for any SONiC use-case, we can think about providing the flexibility in the future.

nmoray replied:
With this approach, we are already providing the flexibility to the user (why to wait for the future :-)). They can have either same or different keys for different features.

The proprietary NOSes have different architectures and they have implemented the feature which can be fitted into their infrastructure. :-) I guess, it is always better if we can add bit of flexibility into any of the designs.

venkatmahalingam commented on src/sonic-py-common/sonic_py_common/security_cipher.py:
```diff
@@ -0,0 +1,120 @@
+import os
+from swsscommon.swsscommon import ConfigDBConnector
+
+class security_cipher:
```
Can we rename this module to "master key manager" or any other better name to align with other popular NOS?

nmoray replied:
I have chosen this name as this module is mainly responsible for securing the cipher. :-)

madhupalu replied:
@nmoray Make sense. However, please rename secuirty_cipher class name to master_key_manager as it is not just securing the cipher and also storing and takes care of lifecycle of it.

venkatmahalingam replied:
@nmoray Are you planning to make changes for the above suggestion?

venkatmahalingam commented on src/sonic-py-common/sonic_py_common/security_cipher.py:
```diff
@@ -0,0 +1,120 @@
+            except PermissionError:
+                syslog.syslog(syslog.LOG_ERR, "__write_passwd_file: Read permission denied: {}".format(self._file_path))
+
+    # Read cipher pass file and return the feature specifc
```
What APIs are exposed from this module so as to use them from features like TACACS+, LDAP..etc

nmoray replied:
Three APIs.

Encrypt passkey
Decrypt passkey
Check if the encrypt_key is enabled for the given feature.

venkatmahalingam replied:
How are these APIs being accessed from hostcfgd for TACACS+?

nmoray replied:
Yeah, I will be updating the existing PRs to incorporate these new changes. But for the clarification, hostcfgd will have to import the security_cipher class, get the instance and call the decrypt_passkey method to get the original passkey in a plaintext format.

madhupalu replied:
@nmoray let's change the class name to master_key_manager and update the HLD accordingly.

venkatmahalingam commented on src/sonic-py-common/sonic_py_common/security_cipher.py:
```diff
@@ -0,0 +1,120 @@
+
+                    #os.chmod(self._file_path, 0o777)
+                    with open(self._file_path, 'w') as file:
+                        file.writelines(lines)
```
Is this 'root' only accessible file?

nmoray replied:
Yes it is.

`root@sonic:/tmp# ls /etc/cipher_pass -l

-rw-r--r-- 1 root root 102 Nov 16 16:44 /etc/cipher_pass`

Yarden-Z replied:
Why is it globally readable?
Shouldn't we restrict this to:
-rw-r-----
At the very least?

nmoray replied:
Yeah, it is doable but will it make much difference? config_db.json is too globally readable, isn't it? :-)

nmoray replied:
@Yarden-Z As requested, updated the access permission from 644 to 640.
```
sonic:~# ls /etc/cipher_pass -l
-rw-r----- 1 root root 91 Feb  1 01:18 /etc/cipher_pass 
```

madhupalu commented on src/sonic-py-common/sonic_py_common/security_cipher.py:
```diff
@@ -0,0 +1,120 @@
+                with open(self._file_path, 'w') as file:
+                    file.writelines("#Auto generated file for storing the encryption passwords\n")
+                    file.writelines("TACPLUS : \nRADIUS : \nLDAP :\n")
+                    #os.chmod(self._file_path, 0o644)
```
why is the line commented, don't you need make the file readonly? if not clean it up.

nmoray replied:
Sure. It is not required.

madhupalu commented on src/sonic-py-common/sonic_py_common/security_cipher.py:
```diff
@@ -0,0 +1,120 @@
+
+A common module for handling the encryption and
+decryption of the feature passkey. It also takes
+care of storying the secure cipher at root 
```
fix the typo - storing

nmoray replied:
Sure

madhupalu commented on src/sonic-py-common/tests/test_security_cipher.py:
```diff
@@ -23,7 +24,8 @@ def test_passkey_encryption(self):
-        with mock.patch("sonic_py_common.security_cipher.ConfigDBConnector", new=ConfigDBConnector):
+        with mock.patch("sonic_py_common.security_cipher.ConfigDBConnector", new=ConfigDBConnector), \
+             mock.patch("builtins.open", mock_open()) as mock_file:
             temp = security_cipher()
```
may need indent code, pl chk

madhupalu replied:
may need indent code, pl chk

madhupalu commented on src/sonic-py-common/tests/test_security_cipher.py:
```diff
@@ -13,7 +13,8 @@
-        with mock.patch("sonic_py_common.security_cipher.ConfigDBConnector", new=ConfigDBConnector):
+        with mock.patch("sonic_py_common.security_cipher.ConfigDBConnector", new=ConfigDBConnector), \
+             mock.patch("builtins.open", mock_open()) as mock_file:
             temp = security_cipher()
```
May need to indent code, pl chk.

madhupalu reviewed (approved):
Changes looks good me except few minor comments. please address it.

nmoray commented:
@venkatmahalingam I have added the AUT too. Can you please approve the PR now? This is an independent module. It doesn't have any impact on any other functionalities. Thus we can merge this into master. I will work on other PRs parallelly to integrate this module into TACACS feature.

nmoray commented:
@venkatmahalingam [PR: TACACS passkey encryption Part - I](https://github.com/sonic-net/sonic-utilities/pull/3027)

venkatmahalingam reviewed (changes requested):
Please address the outstanding comments.

nmoray commented:
> Please address the outstanding comments.

@venkatmahalingam I have replied to the comments. IMO, Flexible design is good approach. Why to restrict user to go on a certain path only. He / she should have multiple options.

avinothcgl commented:
> > Please address the outstanding comments.
> 
> @venkatmahalingam I have replied to the comments. IMO, Flexible design is a good approach. Why to restrict user to go on a certain path only. He / she should have multiple options.

@nmoray @venkatmahalingam giving multiple option is good but also customers who migrates from other well known NOS to SONiC will be using one master key. To accommodate those customers , Can you include one master key option as well.

nmoray commented:
> > > Please address the outstanding comments.
> > 
> > 
> > @venkatmahalingam I have replied to the comments. IMO, Flexible design is a good approach. Why to restrict user to go on a certain path only. He / she should have multiple options.
> 
> @nmoray @venkatmahalingam giving multiple option is good but also customers who migrates from other well known NOS to SONiC will be using one master key. To accommodate those customers , Can you include one master key option as well.

@avinothcgl Using one master key for all the feature is still possible with the current implementation. Just keep same master key for all in cipher_pass file.

For instance, 
"#Auto generated file for storing the encryption passwords"
TACPLUS : <SAME_MASTER_KEY>
RADIUS : <SAME_MASTER_KEY>
LDAP : <SAME_MASTER_KEY>

One more advantage of this design is, same security_cipher module can be extended to other applications as well which need to make use of encryption / decryption. That application can have it's own master_key which will be different from TACPLUS, RADIUS and LDAP.

zhangyanzhao commented:
Reviewers, if you are ok with this PR, please help to approve it. Thanks.

nmoray commented:
@venkatmahalingam As requested, updated the class name as "master_key_mgr".

brholmes1 commented:
This is a great feature! In addition to the TACPLUS PRs already proposed, I would also recommend that the RADIUS keys (partially implemented per this PR) and SNMPv3 user passwords/privs also have the ability to be encrypted to align with industry security standards. If a new issue needs to be opened for tracking of RADIUS/SNMPv3 components let me know and I can submit them.

nmoray commented:
> This is a great feature! In addition to the TACPLUS PRs already proposed, I would also recommend that the RADIUS keys (partially implemented per this PR) and SNMPv3 user passwords/privs also have the ability to be encrypted to align with industry security standards. If a new issue needs to be opened for tracking of RADIUS/SNMPv3 components let me know and I can submit them.

Thanks @brholmes1. IMO, let's first merge the initial infra. Later, we can easily extend the support for other modules. For tracking purpose, please open the new issues in the community.

nmoray commented:
@lguohan Can you please help in approving / merging this PR.

nmoray commented:
Folks please help in merging the PR if there are no more comments. It is open from more than 6 months now. Please help in merging the PR.
cc: @zhangyanzhao @qiluo-msft

qiluo-msft commented on src/sonic-yang-models/yang-models/sonic-system-tacacs.yang:
```diff
@@ -93,7 +93,7 @@ module sonic-system-tacacs {
 
                 leaf passkey {
```
Agree. @nmoray Could you check this comment?

nmoray replied:
Yes. @qiluo-msft @anders-nexthop you are right. I will add it.

nmoray commented on src/sonic-py-common/sonic_py_common/security_cipher.py:
```diff
@@ -0,0 +1,133 @@
+            if not os.path.exists(self._file_path):
+                with open(self._file_path, 'w') as file:
+                    file.writelines("#Auto generated file for storing the encryption passwords\n")
+                    file.writelines("TACPLUS : \nRADIUS : \nLDAP :\n")
```
sure

nmoray commented on src/sonic-py-common/sonic_py_common/security_cipher.py:
```diff
@@ -0,0 +1,133 @@
+                    # Update the password for given feature
+                    lines[self._feature_list.index(feature_type)] = feature_type + ' : ' + passwd + '\n'
+
+                    os.chmod(self._file_path, 0o777)
```
While creating a file, it is having "0o640" permissions only. Thus, while updating / writing the same file, we need a write permission.
Once done with file writing, I am again reverting the permissions to "0o640" to make it readable only.

nmoray commented on src/sonic-py-common/sonic_py_common/security_cipher.py:
```diff
@@ -0,0 +1,133 @@
+    # password
+    def __read_passwd_file(self, feature_type):
+        passwd = None
+        if feature_type == 'NA':
```
This case is just for checking the "NA" as NA is already present in _feature_list. No one is allowed to read with feature_type == NA. This is why, I have added that condition. 
Once this is validated, I am validating whether the input is present in the list or not.

` if feature_type in self._feature_list:
`

nmoray commented on src/sonic-py-common/sonic_py_common/security_cipher.py:
```diff
@@ -0,0 +1,133 @@
+
+        if feature_type in self._feature_list:
+           try:
+               os.chmod(self._file_path, 0o644)
```
Here, we just need to read the content, modification is not required. So provided only the READ permissions only. Previously, we need to explicitly modify the content of the file, so provided R/W permissions.

nmoray commented on src/sonic-py-common/sonic_py_common/security_cipher.py:
```diff
@@ -0,0 +1,133 @@
+    
+    # Encrypt the passkey
+    def encrypt_passkey(self, feature_type, secret, passwd):
+        cmd = [ 'openssl', 'enc', '-aes-128-cbc', '-A',  '-a', '-salt', '-pbkdf2', '-pass', 'pass:' + passwd ]
```
One of our clients is using this symmetric encryption algorithm already. Thus, used the same.

nmoray commented on src/sonic-py-common/sonic_py_common/security_cipher.py:
```diff
@@ -0,0 +1,133 @@
+                return data[key]
+        return False
+
+    def del_cipher_pass(self):
```
Agreed!

nmoray replied:
I will update the method.

mssonicbld commented:
/azp run Azure.sonic-buildimage

nmoray commented:
@qiluo-msft @anders-nexthop I have addressed the comments.

nmoray commented on src/sonic-py-common/tests/test_security_cipher.py:
```diff
@@ -0,0 +1,73 @@
+        ]
+
+
+class TestSecurityCipher(object):
```
Yeah, the plan is to take care of it in the successive commit once this long pending PR has been merged.

nmoray commented:
/azp run Azure.sonic-buildimage

nmoray commented:
@mssonicbld will you please re-trigger the azure build for sonic-buildimage? I guess I do not have the sufficient privileges to do the same.

anders-nexthop reviewed (approved):
Thanks for addressing my review comments!

nmoray commented:
@qiluo-msft Two folks have already approved the PR. Can you please approve too in case there are no more comments.

nmoray commented:
@qiluo-msft can you please help in merging the long pending PR if there are no more comments.

qiluo-msft commented:
/azp run Azure.sonic-buildimage

qiluo-msft commented:
@nmoray Semgrep identified valid issues. Could you check and fix them?

anders-nexthop commented:
> > This is a great feature! In addition to the TACPLUS PRs already proposed, I would also recommend that the RADIUS keys (partially implemented per this PR) and SNMPv3 user passwords/privs also have the ability to be encrypted to align with industry security standards. If a new issue needs to be opened for tracking of RADIUS/SNMPv3 components let me know and I can submit them.
> 
> Thanks @brholmes1. IMO, let's first merge the initial infra. Later, we can easily extend the support for other modules. For tracking purpose, please open the new issues in the community.

I have opened new issues to track these extensions:
SNMP: https://github.com/sonic-net/sonic-buildimage/issues/22277
RADIUS: https://github.com/sonic-net/sonic-buildimage/issues/22276

nmoray commented:
> @nmoray Semgrep identified valid issues. Could you check and fix them?

@qiluo-msft I checked the Semgrep failures. Out of three, two are easy to fix which are related to file permissions. But one is related to shell issue which is bit tricky to fix as we need shell access to execute openssl encryption command. If we set shell=false then the encryption fails.

```
❯❯❱ python.lang.security.audit.subprocess-shell-true.subprocess-shell-true
          Found 'subprocess' function 'Popen' with 'shell=True'. This is dangerous because this call will   
          spawn the command using a shell process. Doing so propagates current shell settings and variables,
          which makes it much easier for a malicious actor to execute commands. Use 'shell=False' instead.  
          Details: https://sg.run/J92w                                                                      
                                                                                                            
           ▶▶┆ Autofix ▶ False
          106┆ proc = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE)
```

anders-nexthop commented:
> > @nmoray Semgrep identified valid issues. Could you check and fix them?
> 
> @qiluo-msft I checked the Semgrep failures. Out of three, two are easy to fix which are related to file permissions. But one is related to shell issue which is bit tricky to fix as we need shell access to execute openssl encryption command. If we set shell=false then the encryption fails.
> 
> ```
> ❯❯❱ python.lang.security.audit.subprocess-shell-true.subprocess-shell-true
>           Found 'subprocess' function 'Popen' with 'shell=True'. This is dangerous because this call will   
>           spawn the command using a shell process. Doing so propagates current shell settings and variables,
>           which makes it much easier for a malicious actor to execute commands. Use 'shell=False' instead.  
>           Details: https://sg.run/J92w                                                                      
>                                                                                                             
>            ▶▶┆ Autofix ▶ False
>           106┆ proc = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE)
> ```

@nmoray can you post the failure?

nmoray commented:
> > > @nmoray Semgrep identified valid issues. Could you check and fix them?
> > 
> > 
> > @qiluo-msft I checked the Semgrep failures. Out of three, two are easy to fix which are related to file permissions. But one is related to shell issue which is bit tricky to fix as we need shell access to execute openssl encryption command. If we set shell=false then the encryption fails.
> > ```
> > ❯❯❱ python.lang.security.audit.subprocess-shell-true.subprocess-shell-true
> >           Found 'subprocess' function 'Popen' with 'shell=True'. This is dangerous because this call will   
> >           spawn the command using a shell process. Doing so propagates current shell settings and variables,
> >           which makes it much easier for a malicious actor to execute commands. Use 'shell=False' instead.  
> >           Details: https://sg.run/J92w                                                                      
> >                                                                                                             
> >            ▶▶┆ Autofix ▶ False
> >           106┆ proc = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE)
> > ```
> 
> @nmoray can you post the failure?

@anders-nexthop If we mark the shell as false then the function fails with below traceback,
    ```
raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'echo U2FsdGVkX18FLV2bmL63fBWvIvp/N/RzX/ulmVRnxHc= | openssl enc -aes-128-cbc -a -d -salt -pbkdf2 -pass pass:test'
```
Thus, I tried to rewrite the logic as below,
```
def decrypt_passkey(encrypted_text: str, password: str) -> str:
    cmd = [
        "openssl", "enc", "-aes-128-cbc", "-a", "-d", "-pbkdf2",
        "-pass", f"pass:{password}"
    ]
    cleaned_input = encrypted_text.strip().encode()
    try:
        result = subprocess.run(
            cmd,
            input=cleaned_input,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            check=True
        )
        return result.stdout.decode().strip()
    except subprocess.CalledProcessError as e:
        print("Decryption failed:", e.stderr.decode())
        return ""
```
But it is also not working, getting below error.
`Decryption failed: error reading input file
`

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

nmoray commented:
@qiluo-msft Will you please help me understand the test failures? Why the  kvmtest-* failed?

anders-nexthop commented:
> @qiluo-msft Will you please help me understand the test failures? Why the kvmtest-* failed? BTW I have fixed the Semgrep errors.

It looks like unrelated failures in test setup code to me, but @qiluo-msft will know more. 

From ElasticTest page:
```
Sanity check failed for vms-kvm-t1-lag on running test_pretest.py|||vms-kvm-t1-lag_399220
```

I looked through and saw a few failures that were ignored (NTP, a bad key `"get_testbed_info\nKeyError: 'vms-kvm-t1-lag'\n"`, and an ACL one that's seemingly always there). I don't see how any of these would be related to your change.

You could try pulling the latest from master branch and pushing that to re-trigger the tests to make sure it's not an intermittent issue.

anders-nexthop commented:
> > @qiluo-msft Will you please help me understand the test failures? Why the kvmtest-* failed? BTW I have fixed the Semgrep errors.
> 
> It looks like unrelated failures in test setup code to me, but @qiluo-msft will know more.
> 
> From ElasticTest page:
> 
> ```
> Sanity check failed for vms-kvm-t1-lag on running test_pretest.py|||vms-kvm-t1-lag_399220
> ```
> 
> I looked through and saw a few failures that were ignored (NTP, a bad key `"get_testbed_info\nKeyError: 'vms-kvm-t1-lag'\n"`, and an ACL one that's seemingly always there). I don't see how any of these would be related to your change.
> 
> You could try pulling the latest from master branch and pushing that to re-trigger the tests to make sure it's not an intermittent issue.

I tried to re-trigger the failed tests manually and got this error:

`TF400813: The user 'aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa' is not authorized to access this resource.`

mssonicbld commented:
/azp run Azure.sonic-buildimage

nmoray commented:
> hat to re-trigger the tests to make sure it's not an

@anders-nexthop you mean just rebase the 

> > > @qiluo-msft Will you please help me understand the test failures? Why the kvmtest-* failed? BTW I have fixed the Semgrep errors.
> > 
> > 
> > It looks like unrelated failures in test setup code to me, but @qiluo-msft will know more.
> > From ElasticTest page:
> > ```
> > Sanity check failed for vms-kvm-t1-lag on running test_pretest.py|||vms-kvm-t1-lag_399220
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I looked through and saw a few failures that were ignored (NTP, a bad key `"get_testbed_info\nKeyError: 'vms-kvm-t1-lag'\n"`, and an ACL one that's seemingly always there). I don't see how any of these would be related to your change.
> > You could try pulling the latest from master branch and pushing that to re-trigger the tests to make sure it's not an intermittent issue.
> 
> I tried to re-trigger the failed tests manually and got this error:
> 
> `TF400813: The user 'aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa' is not authorized to access this resource.`

@anders-nexthop I just rebased my remote branch and  re-pushed it. Let's see if it makes any difference. 
@qiluo-msft BTW, In case these failures are known ones, will you please help in merging the changes.

nmoray commented:
> > hat to re-trigger the tests to make sure it's not an
> 
> @anders-nexthop you mean just rebase the
> 
> > > > @qiluo-msft Will you please help me understand the test failures? Why the kvmtest-* failed? BTW I have fixed the Semgrep errors.
> > > 
> > > 
> > > It looks like unrelated failures in test setup code to me, but @qiluo-msft will know more.
> > > From ElasticTest page:
> > > ```
> > > Sanity check failed for vms-kvm-t1-lag on running test_pretest.py|||vms-kvm-t1-lag_399220
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > I looked through and saw a few failures that were ignored (NTP, a bad key `"get_testbed_info\nKeyError: 'vms-kvm-t1-lag'\n"`, and an ACL one that's seemingly always there). I don't see how any of these would be related to your change.
> > > You could try pulling the latest from master branch and pushing that to re-trigger the tests to make sure it's not an intermittent issue.
> > 
> > 
> > I tried to re-trigger the failed tests manually and got this error:
> > `TF400813: The user 'aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa' is not authorized to access this resource.`
> 
> @anders-nexthop I just rebased my remote branch and re-pushed it. Let's see if it makes any difference. @qiluo-msft BTW, In case these failures are known ones, will you please help in merging the changes.

@qiluo-msft The checks are still failing. Please let us know what needs to be done to fix these failures.

anders-nexthop commented on src/sonic-py-common/tests/test_security_cipher.py:
```diff
@@ -0,0 +1,73 @@
+                encrypt = temp.encrypt_passkey("RADIUS", "passkey2", "TEST2")
+
+            # Use patch to replace the built-in 'open' function with a mock
+            #with mock.patch("{}.open".format(BUILTINS), mock.mock_open(read_data=EXPECTED_PASSWD)) as mock_file:
```
still needed?

nmoray commented:
@qiluo-msft Please re-approve the PR. I pushed one more commit for removing commented code.

nmoray commented:
@qiluo-msft can you please re-approve and merge the PR. All the checks have been passed now.

qiluo-msft commented on src/sonic-yang-models/yang-models/sonic-system-tacacs.yang:
```diff
@@ -131,9 +131,14 @@ module sonic-system-tacacs {
                 }
 
+		leaf key_encrypt {
+		    type boolean;
```
>boolean [](http://example.com/codeflow?start=11&length=7)

Please add a default value (false). This is needed for backward compatible.

nmoray commented on src/sonic-py-common/sonic_py_common/security_cipher.py:
```diff
@@ -0,0 +1,184 @@
+
+    def __init__(self):
+        if not self._initialized:
+            self._file_path = "/etc/cipher_pass"
```
@qiluo-msft As per https://github.com/sonic-net/SONiC/blob/master/doc/ztp/SONiC-config-setup.md#222-config-migration
the backup and restore of cipher_pass file will automatically be taken care, right?

nmoray commented on src/sonic-yang-models/yang-models/sonic-system-tacacs.yang:
```diff
@@ -131,9 +131,14 @@ module sonic-system-tacacs {
                     default 5;
                 }
 
+		leaf key_encrypt {
```
You mean, replace the tabs with spaces for the entire leaf section?

nmoray replied:
The link that you shared is not accessible.
